### ISSUE TITLE
perlop / list assignment documentation

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#9820]
Migrated from [rt.perl.org#68312](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=68312) (status was 'rejected')

Searchable as RT68312$

**Comment by p5pRT** [comment]
### From @ig3

#### Created by ian@alula.local



Here is a patch to perlop\.pod which\, I think\, clarifies the
description of the assignment operator a little\.

<details><summary>Perl Info</summary>

```

Flags:
    category=docs
    severity=low

Site configuration information for perl 5.11.0:

Configured by ian at Sun Mar 29 08:00:08 NZDT 2009.

Summary of my perl5 (revision 5 version 11 subversion 0) configuration:
  Commit id:
  Platform:
    osname=linux, osvers=2.6.18-92.el5, archname=i686-linux
    uname='linux alula.local 2.6.18-92.el5 #1 smp tue jun 10 18:49:47 
edt 2008 i686 i686 i386 gnulinux '
    config_args='-DDEBUGGING'
    hint=previous, useposix=true, d_sigaction=define
    useithreads=undef, usemultiplicity=undef
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=undef, use64bitall=undef, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-DDEBUGGING -fno-strict-aliasing -pipe 
-fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE 
-D_FILE_OFFSET_BITS=64',
    optimize='-g',
    cppflags='-DDEBUGGING -fno-strict-aliasing -pipe -fstack-protector 
-I/usr/local/include -DDEBUGGING -fno-strict-aliasing -pipe 
-fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE 
-D_FILE_OFFSET_BITS=64 -DDEBUGGING -fno-strict-aliasing -pipe 
-fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE 
-D_FILE_OFFSET_BITS=64'
    ccversion='', gccversion='4.1.2 20071124 (Red Hat 4.1.2-42)', 
gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', 
lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -fstack-protector -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -lgdbm -ldb -ldl -lm -lcrypt -lutil -lc
    perllibs=-lnsl -ldl -lm -lcrypt -lutil -lc
    libc=/lib/libc-2.5.so, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.5'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -g -L/usr/local/lib 
-fstack-protector'

Locally applied patches:
    PERL_GIT_UNPUSHED_COMMITS           /* do not remove this line */
    PERL_GIT_UNCOMMITTED_CHANGES        /* do not remove this line */


@INC for perl 5.11.0:
    /home/ian/localperl/lib/site_perl/5.11.0/i686-linux
    /home/ian/localperl/lib/site_perl/5.11.0
    /home/ian/localperl/lib/5.11.0/i686-linux
    /home/ian/localperl/lib/5.11.0
    .


Environment for perl 5.11.0:
    HOME=/home/ian
    LANG=en_NZ.UTF-8
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    
PATH=/usr/local/java/jdk/bin:/usr/local/java/ant/bin:/usr/kerberos/bin:/usr/local/bin:/bin:/usr/bin:/usr/local/Trolltech/Qt-4.4.3/bin:/home/ian/bin:/sbin
    PERLDOC_PAGER=less
    PERL_BADLANG (unset)
    SHELL=/bin/bash


```
</details>

**Comment by p5pRT** [comment]
### From @ig3

<details><summary>0001-Clarified-description-of-the-assignment-operator.patch</summary>

```diff
From 165a188b1d79244c58e9ed0d4be753b6ee674f7d Mon Sep 17 00:00:00 2001
From: Ian Goodacre <Ian.Goodacre@xtra.co.nz>
Date: Sat, 8 Aug 2009 11:10:17 +1200
Subject: [PATCH] Clarified description of the assignment operator.

---
 pod/perlop.pod |   27 ++++++++++++++++++++-------
 1 files changed, 20 insertions(+), 7 deletions(-)

diff --git a/pod/perlop.pod b/pod/perlop.pod
index e677430..b9ec357 100644
--- a/pod/perlop.pod
+++ b/pod/perlop.pod
@@ -731,7 +731,12 @@ X<%=> X<^=> X<x=>
 
 "=" is the ordinary assignment operator.
 
-Assignment operators work as in C.  That is,
+There are two forms of the assignment operator: scalar assignment and list
+assignment.
+
+If the left operand is a scalar lvalue then the assignment operator
+is a scalar assignment operator. Scalar assignment operators work
+as in C. That is,
 
     $a += 2;
 
@@ -740,8 +745,8 @@ is equivalent to
     $a = $a + 2;
 
 although without duplicating any side effects that dereferencing the lvalue
-might trigger, such as from tie().  Other assignment operators work similarly.
-The following are recognized:
+might trigger, such as from tie().  Other scalar assignment operators work
+similarly.  The following are recognized:
 
     **=    +=    *=    &=    <<=    &&=
            -=    /=    |=    >>=    ||=
@@ -767,10 +772,18 @@ is equivalent to
     $a += 2;
     $a *= 3;
 
-Similarly, a list assignment in list context produces the list of
-lvalues assigned to, and a list assignment in scalar context returns
-the number of elements produced by the expression on the right hand
-side of the assignment.
+If the left operand of the assignment operator is a list then the operator
+is a list assignment operator. In this case, each element of the list
+must be an lvalue or C<undef>. The right operand is evaluated in list context
+and elements of the resulting list are assigned to elements of the list on the
+left hand side. See L<List value constructors> in L<perldata> for examples.
+
+In scalar context, the list assignment operator returns the number of 
+elements produced by the expression on the right hand side of the 
+assignment.
+
+In list context, the list assignment operator returns the list of
+lvalues assigned to.
 
 =head2 Comma Operator
 X<comma> X<operator, comma> X<,>
-- 
1.6.1


```
</details>

**Comment by p5pRT** [comment]
### From Eirik-Berg.Hanssen@allverden.no

"ian\.goodacre@&#8203;xtra\.co\.nz \(via RT\)" \<perlbug\-followup@&#8203;perl\.org> writes&#8203;:

> \+If the left operand of the assignment operator is a list then the operator

&nbsp;  Please don't call the left operand a "list"\.  That is not just
misleading\, but also risks propagating rather tough misconceptions\.

&nbsp;  Consider&#8203;: "is an array\, a hash\, a slice\, or enclosed in parentheses"\.

&nbsp;  Consider&#8203;: In the following\, is f a list? a scalar? both?

use 5\.010;
use strict;
use warnings;
my $x=1;
my $y=2;
sub f \(\) :lvalue \{
&nbsp;  $x\, $y
\}
say f;
f=4;
say f;
f=\(5\,6\);
say f
\_\_END\_\_
Useless use of a constant in void context at \- line 12\.
12
14
16


Eirik
\-\- 
Flon's Law&#8203;:
&nbsp;        There is not now\, and never will be\, a language in which it is
&nbsp;        the least bit difficult to write bad programs\.

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From Eirik-Berg.Hanssen@allverden.no

Eirik Berg Hanssen \<Eirik\-Berg\.Hanssen@&#8203;allverden\.no> writes&#8203;:

>   Consider&#8203;: In the following\, is f a list? a scalar? both?
>
> use 5\.010;
> use strict;
> use warnings;
> my $x=1;
> my $y=2;
> sub f \(\) :lvalue \{
>   $x\, $y
> \}
> say f;
> f=4;
> say f;
> f=\(5\,6\);
> say f
> \_\_END\_\_
> Useless use of a constant in void context at \- line 12\.
> 12
> 14
> 16

&nbsp;  Come to think of it\, I could do that one even clearer&#8203;:

use 5\.010;
use strict;
use warnings;
my $x=1;
my $y=2;
sub f \(\) :lvalue \{
&nbsp;  $x\, $y
\}
say f;
say f=4;
say f;
say f=\(5\,6\);
say f;
\_\_END\_\_
Useless use of a constant in void context at \- line 12\.
12
4
14
6
16


Eirik
\-\- 
Ceci n'est pas une \.sig\.

**Comment by p5pRT** [comment]
### From @ig3

Eirik Berg Hanssen via RT wrote&#8203;:
> "ian\.goodacre@&#8203;xtra\.co\.nz \(via RT\)" \<perlbug\-followup@&#8203;perl\.org> writes&#8203;:
>
>   
>> \+If the left operand of the assignment operator is a list then the operator
>>     
>
>   Please don't call the left operand a "list"\.  That is not just
> misleading\, but also risks propagating rather tough misconceptions\.
>
>   Consider&#8203;: "is an array\, a hash\, a slice\, or enclosed in parentheses"\.
>
>   Consider&#8203;: In the following\, is f a list? a scalar? both?
>
> use 5\.010;
> use strict;
> use warnings;
> my $x=1;
> my $y=2;
> sub f \(\) :lvalue \{
>   $x\, $y
> \}
> say f;
> f=4;
> say f;
> f=\(5\,6\);
> say f
> \_\_END\_\_
> Useless use of a constant in void context at \- line 12\.
> 12
> 14
> 16
>
>
> Eirik
>   

Despite having submitted the patch\, because I think the description 
should be improved\, I don't know enough to decide how to describe the 
assignment operator\.

I would be happy to submit a revised patch incorporating your \(Eirik's\) 
suggestion if that helps\. I think it is a significant improvement over 
what I wrote \(Eirik\+\+\)\. I would also be happy\, and not at all offended\, 
if someone wrote something entirely different and more correct\. I was 
motivated primarily because the terms "list assignment" and "scalar 
assignment" are used without being defined\. Any revision which 
accurately defines them would be great\.

I am curious why you \(Eirik\) say "enclosed in parentheses" rather than 
"a list"\. This\, no doubt\, results from my abiding confusion as to 
whether or when parentheses enclosing an expression constitutes a list 
and when they are merely affecting precedence or make no difference at 
all\. Are there cases where a list as a left operand would not cause the 
assignment to be a list assignment? Are there cases where a left operand 
being enclosed in parentheses would not be a list? I would be very 
interested to see example\, to help my learning\.

I am uncertain whether it makes more sense to say that the assignment 
operator provides context to its left operand or its left operand 
provides context to the assignment operator\. In the statement 
'f=\(5\,6\);'\, it appears that both operands are evaluated in scalar 
context\. Yet it doesn't make much sense to me to say that the assignment 
operator provides scalar context to its left operand given that the 
assignment operator is a scalar or list assignment operator depending on 
the nature of its left operand \- i\.e\. the left operand seems to 
determine the nature of the assignment operator\, not vice\-versa\. In the 
statement '\($x\,$y\)=\(5\,6\);' the left operand is not evaluated in scalar 
context \(in which case it would not be a list and we would have scalar 
assignment of 6 to $y with $x and 5 being evaluated in void context\)\. I 
am very confused\.

Regards\,
Ian Goodacre

**Comment by p5pRT** [comment]
### From @davidnicol

On Mon\, Aug 10\, 2009 at 7&#8203;:44 AM\, Ian Goodacre\<Ian\.Goodacre@&#8203;xtra\.co\.nz>

> I am curious why you \(Eirik\) say "enclosed in parentheses" rather than "a
> list"\.

As I understand it\, there is a weak taboo against calling an "array"
data structure a "list" because "list" is supposed to be strictly
about syntax\. I do not agree with this convention myself and often
break it because I think "list" is a perfectly fine word for an
ordered collection\. That's one of the hiss\-fights you've stepped into\.

> This\, no doubt\, results from my abiding confusion as to whether or
> when parentheses enclosing an expression constitutes a list and when they
> are merely affecting precedence or make no difference at all\. Are there
> cases where a list as a left operand would not cause the assignment to be a
> list assignment? Are there cases where a left operand being enclosed in
> parentheses would not be a list? I would be very interested to see example\,
> to help my learning\.

Me too\.  I think a chart of various examples and what happens with
them would be very much in order near the assignment operator\.

> I am uncertain whether it makes more sense to say that the assignment
> operator provides context to its left operand or its left operand provides
> context to the assignment operator\. In the statement 'f=\(5\,6\);'\, it appears
> that both operands are evaluated in scalar context\.

lvalue subroutines are always scalar\.  "list" syntax\, in scalar
context\, causes the commas to be interpreted as "the comma operator"
rather than the "list element separator operator\." Fat arrows\, too\.

catnip&#8203;:\~\# perl \-wle 'my $x = \(5\,6\); print $x'
Useless use of a constant in void context at \-e line 1\.
6
catnip&#8203;:\~\# perl \-wle 'my $x = @&#8203;y = \(5\,6\); print $x'
Name "main&#8203;::y" used only once&#8203;: possible typo at \-e line 1\.
2

In the first\, the list is parsed as a list\. In the second\, the list is
parsed as an array\.


> Yet it doesn't make much
> sense to me to say that the assignment operator provides scalar context to
> its left operand given that the assignment operator is a scalar or list
> assignment operator depending on the nature of its left operand \- i\.e\. the
> left operand seems to determine the nature of the assignment operator\, not
> vice\-versa\. In the statement '\($x\,$y\)=\(5\,6\);' the left operand is not
> evaluated in scalar context \(in which case it would not be a list and we
> would have scalar assignment of 6 to $y with $x and 5 being evaluated in
> void context\)\. I am very confused\.

Context is provided by the left side of an assignment\. lvalue
subroutines when used as lvalues are always scalar\.

As l\-values\, parentheses make lists\. As r\-values\, they make lists in
list context or guide precedence in scalar context\.  This can be
confusing&#8203;:

catnip&#8203;:\~\# perl \-wle 'my $x = \(\) = \(5\,6\); print $x'
2

We might say that the empty list is parsed as an array\, then evaluated
in scalar context\.


> Regards\,
> Ian Goodacre
>



\-\- 
Today in 1680 Taos besieged Santa Fe
http&#8203;://www\.amazon\.com/exec/obidos/ISBN=1574160648/tipjartransactioA

**Comment by p5pRT** [comment]
### From Eirik-Berg.Hanssen@allverden.no

David Nicol \<davidnicol@&#8203;gmail\.com> writes&#8203;:

> On Mon\, Aug 10\, 2009 at 7&#8203;:44 AM\, Ian Goodacre\<Ian\.Goodacre@&#8203;xtra\.co\.nz>
>
>> I am curious why you \(Eirik\) say "enclosed in parentheses" rather than "a
>> list"\.

&nbsp;  Because parentheses do not make a list\, except in a certain
interpretation of the left hand operand of either the assignment
operator or the \(in list context\) repetition operator\.

&nbsp;  Nowhere else\.

&nbsp;  \(Unless I'm forgetting yet another special case\.\)


> lvalue subroutines are always scalar\.

&nbsp;  Really now?

use 5\.010;
use strict;
use warnings;
my $x=1;
my $y=2;
sub f \(\) :lvalue \{
&nbsp;  $x\, $y
\}
say f;
say\( \(f\)=\(4\) \);
say\( \(f\)\.\(4\) \);
say f;
say\( \(f\)=\(5\,6\) \);
say\( \(f\)\.\(5\,6\) \);
say f;
\_\_END\_\_
Useless use of a constant in void context at \- line 14\.
Use of uninitialized value in say at \- line 10\.
Use of uninitialized value in concatenation \(\.\) or string at \- line 11\.
Use of uninitialized value in say at \- line 12\.
12
4
4
4
56
66
56

&nbsp;  Not only does say\(\) evaluate f in list context\, list assignment does
too\!  \(When\, as here\, the left operand of the assignment operator is
enclosed in parentheses\, it is a list assignment\.\)


&nbsp;  In a nutshell&#8203;: The lvalueness of f does not determine whether it
produces a scalar or a list\.  What matters is context\.

&nbsp;  Always context\.

&nbsp;  And the list context seems to be provided by the parentheses\, but is
really provided by the list assignment operator\.  In the examples with
the concatenation operator\, it is called in scalar context – even with
parentheses\!


>> Yet it doesn't make much
>> sense to me to say that the assignment operator provides scalar context to
>> its left operand given that the assignment operator is a scalar or list
>> assignment operator depending on the nature of its left operand \- i\.e\. the
>> left operand seems to determine the nature of the assignment operator\, not
>> vice\-versa\. In the statement '\($x\,$y\)=\(5\,6\);' the left operand is not
>> evaluated in scalar context \(in which case it would not be a list and we
>> would have scalar assignment of 6 to $y with $x and 5 being evaluated in
>> void context\)\. I am very confused\.

&nbsp;  This example is a list assignment because of the parentheses\, and
therefore the operands are evaluated in list context\.


> Context is provided by the left side of an assignment\.

&nbsp;  By the syntax of the assignment\, actually\.  The context is
propagated into the left side operand as well \.\.\.


> As r\-values\, they make lists in
> list context or guide precedence in scalar context\.  This can be
> confusing&#8203;:

&nbsp;  No\, they only guide precedence either way\.

use 5\.010;
use strict;
use warnings;
say for \(1\, 0\)  \# scalar context
&nbsp;     or \(2\, 3\); \# list context
say for  1\, 0   \# still scalar context
&nbsp;     or  2\, 3;  \# still list context
\_\_END\_\_
2
3
2
3

&nbsp;  Except for scoping considerations\, you could use do\{ \} blocks
instead of parentheses – except where the parentheses are syntactic\.
Which aside from straightforwardly "different" stuff like function
call syntax\, prototype declarations\, and regexen\, would seem to be
list assignment and list repetition\.  \(Again\, I may be forgetting some
other special case\.\)

use 5\.010;
use strict;
use warnings;
my $x=1;

eval '\($x\) = 2' or say $@&#8203;;
eval 'do\{ $x \} = 2' or say $@&#8203;;
say "List repetition&#8203;: "\, join "\-"\, \($x\) x 3;
say "Scalar repetition&#8203;: "\, join "\-"\, do\{$x\} x 3;
\_\_END\_\_
Can't modify do block in scalar assignment at \(eval 2\) line 2\, at EOF

List repetition&#8203;: 2\-2\-2
Scalar repetition&#8203;: 222


&nbsp;  Bottom line&#8203;: There is a whole lot of confusion surrounding lists in
Perl\.  Don't encourage it\.


Eirik
\-\- 
Some people have told me they don't think a fat penguin really embodies the
grace of Linux\, which just tells me they have never seen a angry penguin
charging at them in excess of 100mph\.  They'd be a lot more careful about
what they say if they had\.  \-\- Linus Torvalds\, announcing Linux v2\.0

**Comment by p5pRT** [comment]
### From @davidnicol

On Mon\, Aug 10\, 2009 at 11&#8203;:20 AM\, Eirik Berg Hanssen\<Eirik\-
>> lvalue subroutines are always scalar\.
>
>  Really now?

and by "always scalar" I mean they are like $x not like @&#8203;x when
getting used as l\-values\.  When not being l\-values\, lvalue subroutines
are perfectly normal\. Where "perfectly normal" means "like any other
perl subroutine\."  Sorry for any confusion \-\-  "always"  was poor word
choice\.


\-\- 
Today in 1680 Taos besieged Santa Fe
http&#8203;://www\.amazon\.com/exec/obidos/ISBN=1574160648/tipjartransactioA

**Comment by p5pRT** [comment]
### From Eirik-Berg.Hanssen@allverden.no

David Nicol \<davidnicol@&#8203;gmail\.com> writes&#8203;:

> On Mon\, Aug 10\, 2009 at 11&#8203;:20 AM\, Eirik Berg Hanssen\<Eirik\-
>>> lvalue subroutines are always scalar\.
>>
>>  Really now?
>
> and by "always scalar" I mean they are like $x not like @&#8203;x when
> getting used as l\-values\.  When not being l\-values\, lvalue subroutines
> are perfectly normal\. Where "perfectly normal" means "like any other
> perl subroutine\."  Sorry for any confusion \-\-  "always"  was poor word
> choice\.

&nbsp;  Better\, but still not quite right\.  lvalue subroutines are unlike
either of those when used in list assignment&#8203;:

use 5\.010;
use strict;
use warnings;

my\($x\, $y\, $f\, @&#8203;f\, $g\, @&#8203;h\);
sub f \(\) :lvalue \{ $x\, $y \}

say "How many elements do they gobble up?";
\($f\, $g\, @&#8203;h\) = 0\.\.10;
say " A scalar&#8203;: "\, $g // "all";
\(f\, $g\, @&#8203;h\) = 0\.\.10;
say " This lvalue subroutine&#8203;: "\, $g // "all";
\(@&#8203;f\, $g\, @&#8203;h\) = 0\.\.10;
say " An array&#8203;: "\, $g // "all";
\_\_END\_\_
How many elements do they gobble up?
 A scalar&#8203;: 1
 This lvalue subroutine&#8203;: 2
 An array&#8203;: all


&nbsp;  Yeah\, it's crazy\.  Crazy powerful\. :\)  But also crazy complex\. :\-\\


&nbsp;  See how much simpler it is to note that assignment is list
assignment if the left hand operand is an array\, a hash\, a slice\, or
enclosed in parentheses?


Eirik
\-\- 
Live and let learn\, that's my policy\.
&nbsp;            \-\-Larry Wall

**Comment by p5pRT** [comment]
### From @davidnicol

On Mon\, Aug 10\, 2009 at 3&#8203;:18 PM\, Eirik Berg
Hanssen\<Eirik\-Berg\.Hanssen@&#8203;allverden\.no> wrote&#8203;:
> my\($x\, $y\, $f\, @&#8203;f\, $g\, @&#8203;h\);
> sub f \(\) :lvalue \{ $x\, $y \}
>
> \(f\, $g\, @&#8203;h\) = 0\.\.10;
> say " This lvalue subroutine&#8203;: "\, $g // "all";
> \_\_END\_\_
>  This lvalue subroutine&#8203;: 2
>
>  Yeah\, it's crazy\.  Crazy powerful\. :\)  But also crazy complex\. :\-\\

I would classify that gobbling of two list elements as a bug\, as it
isn't consistent with anything else\. Was it intended\, or merely
justified? And do $x and $y get loaded with 0 and 1? If so that would
mean that the whole "l\-value subs are scalar context" thing is false
in the face of overwhelming syntax\, so

&nbsp;   f = \(5\,6\); \# sets $y to 6\, leaves $x alone
&nbsp;   \(f\) = \(5\,6\); \# sets $x to 5 and $y to 6

would be right\.  Which is intensely impressive\.

\-\-
"Galileo never took Tycho's system seriously\, as can be seen in his
correspondence\, regarding it as an inadequate and physically
unsatisfactory compromise\." \-\- Wikipedia

**Comment by p5pRT** [comment]
### From @ig3

Eirik Berg Hanssen wrote&#8203;:
> See how much simpler it is to note that assignment is list
> assignment if the left hand operand is an array\, a hash\, a slice\, or
> enclosed in parentheses?
>
>   
I am thinking that should be&#8203;: an array\, a hash\, an array or hash slice\, 
or enclosed in parentheses\.

Because trying to assign to a list slice produces a compile time error 
on a scalar assignment&#8203;:

my \($x\,$y\,$z\,$h\,$i\,$j\) = \(1\.\.6\);
sub f :lvalue \{ $x\, $y\, $z \}

say f\,$h\,$i\,$j;
\(f\,$h\,$i\,$j\)\[1\.\.4\] = \('a'\.\.'z'\);
say f\,$h\,$i\,$j;
\_\_END\_\_
Can't modify list slice in scalar assignment at \./test\.pl line 10\, near "\);"
Execution of \./test\.pl aborted due to compilation errors\.

Adding parentheses makes it a list assignment\, but it still doesn't compile&#8203;:

my \($x\,$y\,$z\,$h\,$i\,$j\) = \(1\.\.6\);
sub f :lvalue \{ $x\, $y\, $z \}

say f\,$h\,$i\,$j;
\(\(f\,$h\,$i\,$j\)\[1\.\.4\]\) = \('a'\.\.'z'\);
say f\,$h\,$i\,$j;
\_\_END\_\_
Can't modify list slice in list assignment at \./test\.pl line 10\, near "\);"
Execution of \./test\.pl aborted due to compilation errors\.

I guess the compile error is because list slices don't return lvalues\.


Eirik Berg Hanssen wrote&#8203;:
> Because parentheses do not make a list\, except in a certain
> interpretation of the left hand operand of either the assignment
> operator or the \(in list context\) repetition operator\.
>
>   Nowhere else\.
>
>   \(Unless I'm forgetting yet another special case\.\)
>
>   
None the less\, it seems a list slice is a third special case where 
parentheses make a list\.

Ian Goodacre

**Comment by p5pRT** [comment]
### From Eirik-Berg.Hanssen@allverden.no

Ian Goodacre \<Ian\.Goodacre@&#8203;xtra\.co\.nz> writes&#8203;:

> Eirik Berg Hanssen wrote&#8203;:
>> See how much simpler it is to note that assignment is list
>> assignment if the left hand operand is an array\, a hash\, a slice\, or
>> enclosed in parentheses?
>>
>>   
> I am thinking that should be&#8203;: an array\, a hash\, an array or hash
> slice\, or enclosed in parentheses\.
>
> Because trying to assign to a list slice produces a compile time error
> on a scalar assignment&#8203;:

> Adding parentheses makes it a list assignment\, but it still doesn't compile&#8203;:

> I guess the compile error is because list slices don't return lvalues\.

&nbsp;  I was not aware of that\, but it does rather make sense\. :\)

> Eirik Berg Hanssen wrote&#8203;:
>> Because parentheses do not make a list\, except in a certain
>> interpretation of the left hand operand of either the assignment
>> operator or the \(in list context\) repetition operator\.
>>
>>   Nowhere else\.
>>
>>   \(Unless I'm forgetting yet another special case\.\)
>>
>>   
> None the less\, it seems a list slice is a third special case where
> parentheses make a list\.

&nbsp;  Right again\[\*\]\.  Good catches both\. :\)


\[\*\]&#8203;: To the extent that the interpretation to which I was referring is
&nbsp;     "right"\.  Personally\, I'd rather say that the syntax of these
&nbsp;     three operations provides list context to the expression between
&nbsp;     those parentheses\.  But that may just be a knee\-jerk reaction
&nbsp;     against "list anything" that isn't operations or context – the
&nbsp;     only times we \_need\_ the concept of lists\. ;\-\)

Eirik
\-\- 
Duck typing&#8203;: if it walks like a duck\, and talks like a duck\, then you'd
better duck\, because the person drawing the gun is probably not an artist\.

&nbsp;    \-\- Larry Wall

**Comment by p5pRT** [comment]
### From @ig3

Thanks for all the comments and examples\. I have learned a lot\.\.\.

What do you thing of this revised patch?

Ian Goodacre

**Comment by p5pRT** [comment]
### From @ig3

<details><summary>0001-Revised-assignment-operator-docs.patch</summary>

```diff
From e3b7da907b3774fffa3ffeb3b715e914b4612faf Mon Sep 17 00:00:00 2001
From: Ian Goodacre <Ian.Goodacre@xtra.co.nz>
Date: Tue, 11 Aug 2009 22:30:15 +1200
Subject: [PATCH] Revised assignment operator docs

Elaborated description of list assignment operator, providing a description
of left operands that cause an assignment operator to be a list assignment
operator, and reorganized section.
---
 pod/perlop.pod |   53 ++++++++++++++++++++++++++++++++++++-----------------
 1 files changed, 36 insertions(+), 17 deletions(-)

diff --git a/pod/perlop.pod b/pod/perlop.pod
index e677430..bcb5ad4 100644
--- a/pod/perlop.pod
+++ b/pod/perlop.pod
@@ -729,9 +729,26 @@ X<assignment> X<operator, assignment> X<=> X<**=> X<+=> X<*=> X<&=>
 X<<< <<= >>> X<&&=> X<-=> X</=> X<|=> X<<< >>= >>> X<||=> X<//=> X<.=>
 X<%=> X<^=> X<x=>
 
-"=" is the ordinary assignment operator.
+"=" is the ordinary assignment operator.  There are two forms of this
+operator: list assignment and scalar assignment.  If the left operand
+is an array, a hash, an array or hash slice or enclosed in parentheses,
+then the assignment is a list assignment.  Otherwise it is a scalar assignment.
 
-Assignment operators work as in C.  That is,
+There are other assignment operators, as in C. The following are recognized:
+
+    **=    +=    *=    &=    <<=    &&=
+           -=    /=    |=    >>=    ||=
+           .=    %=    ^=           //=
+                 x=
+
+These other assignment operators are all scalar assignment operators.
+Although they are grouped by family, they all have the precedence
+of assignment.
+
+The scalar assignment operators provide scalar context to both their operands.
+The left operand must be an lvalue (meaning you can assign to it).
+
+The other assignment operators work as in C. That is,
 
     $a += 2;
 
@@ -740,18 +757,9 @@ is equivalent to
     $a = $a + 2;
 
 although without duplicating any side effects that dereferencing the lvalue
-might trigger, such as from tie().  Other assignment operators work similarly.
-The following are recognized:
-
-    **=    +=    *=    &=    <<=    &&=
-           -=    /=    |=    >>=    ||=
-           .=    %=    ^=           //=
-                 x=
+might trigger, such as from tie(). 
 
-Although these are grouped by family, they all have the precedence
-of assignment.
-
-Unlike in C, the scalar assignment operator produces a valid lvalue.
+Unlike in C, the scalar assignment operators produce a valid lvalue.
 Modifying an assignment is equivalent to doing the assignment and
 then modifying the variable that was assigned to.  This is useful
 for modifying a copy of something, like this:
@@ -767,10 +775,21 @@ is equivalent to
     $a += 2;
     $a *= 3;
 
-Similarly, a list assignment in list context produces the list of
-lvalues assigned to, and a list assignment in scalar context returns
-the number of elements produced by the expression on the right hand
-side of the assignment.
+The list assignment operator provides list context to both its operands. 
+
+If the left operand of the list assignment operator is enclosed in parentheses,
+each element of the enclosed list must be an lvalue or C<undef>.
+
+In scalar context, the list assignment operator returns the number of 
+elements produced by the expression on the right hand side of the 
+assignment.
+
+In list context, the list assignment operator returns the list of
+lvalues assigned to.
+
+See L<List value constructors> in L<perldata> for examples of list assignment.
+
+
 
 =head2 Comma Operator
 X<comma> X<operator, comma> X<,>
-- 
1.6.1


```
</details>

**Comment by p5pRT** [comment]
### From Eirik-Berg.Hanssen@allverden.no

Ian Goodacre \<Ian\.Goodacre@&#8203;xtra\.co\.nz> writes&#8203;:

> Thanks for all the comments and examples\. I have learned a lot\.\.\.

&nbsp;  You're welcome\.  It's nice when one's rants are appreciated\. :\)

> What do you thing of this revised patch?

&nbsp;  It is not my call\, but since I've involved myself in this request\,
I'll continue speaking my mind \.\.\.

&nbsp;  It looks good to me\.  Certainly a significant improvement on the
docs of today\.  I recommend it\.


Eirik\, with thumbs up
\-\- 
All the books you read and courses you take in college are just there to help
you get familiar with the brushes and the canvas\.  They'll help keep you from
making brown blobs\, but you'll never make a masterpiece by the book\.
&nbsp;  \-\- Robert Simpson\, mail\-archive\.com/sqlite\-users%40sqlite\.org/msg14592\.html

**Comment by p5pRT** [comment]
### From @davidnicol

I like "list assignment in scalar context" as that is exactly what
"the empty list counting trick" is\.


> \+In scalar context\, the list assignment operator returns the number of
> \+elements produced by the expression on the right hand side of the
> \+assignment\.
> \+
> \+In list context\, the list assignment operator returns the list of
> \+lvalues assigned to\.

but that is how lists are in scalar context&#8203;: this behavior is not
peculiar to the assignment operator\, so these two points might be
better combined&#8203;:

&nbsp;     As an r\-value\, the list assignment operator passes its own
r\-value through\. List assignments can be chained\, and in a scalar
context the result of a list assignment is the number of elements
produced by the expression on the right hand side\.

&nbsp;     @&#8203;XmlThings = m\{\<thing>\(\.\+\)\</thing>\}gs or do\{
&nbsp;         warn "no things"; next
&nbsp;     \};

**Comment by p5pRT** [comment]
### From Eirik-Berg.Hanssen@allverden.no

David Nicol \<davidnicol@&#8203;gmail\.com> writes&#8203;:

> that is how lists are in scalar context

&nbsp;  I feel another rant coming up \.\.\. one of the old ones&#8203;:

&nbsp;  Lists are never in scalar context\.


&nbsp;  List assignment may be\, list slices may be\, but lists may not\.

&nbsp;  \(List repetitions may not be in scalar context either\, surpisingly
enough\.  It is an operation that only exists in list context\.\)


>      As an r\-value\, the list assignment operator passes its own
> r\-value through\.

&nbsp;  What is "its own r\-value"?  If you mean to say the "elements
produced by the expression on the right hand side of the assignment"\,
it is incorrect\.

&nbsp;  \(And if you meant something else\, I found it less clear than what
the patch had\.\)


use 5\.010;
use strict;
use warnings;
my @&#8203;x = my \($x\, undef\, $y\) = 1\.\.10;
say $\_||'undef' for @&#8203;x;
\_\_END\_\_
1
2
3

&nbsp;  3 elements\, not 10\.  Specifically\, the three elements assigned to\.


&nbsp;  Aside&#8203;:

&nbsp;  And none of them undef either\.  In fact\, this is one \.\.\. rather
\_special\_ \.\.\. corner case&#8203;:

use 5\.010;
use strict;
use warnings;
for my $u \(undef\) \{
&nbsp;  my @&#8203;x = \($u\) = 1\.\.10;
&nbsp;  say $\_||'undef' for @&#8203;x\, $u;
\}
\_\_END\_\_
1
undef

&nbsp;  The assignment to \($u\) evaluates to a list of one element\, the value
of which differs from the value of $u\, which remains the same both
before and after the assignment\.  Ah\, the magic of undef\. :\)


Eirik
\-\- 
\-\-\-\-\-\-\-\- Rule of Feline Frustration&#8203;: \-\-\-\-\-\-\-\* I bet you never noticed that
When your cat has fallen asleep on your lap | "homeowner" has the word    
and looks utterly content and adorable\, you | "meow" in the middle of it\. 
will suddenly have to go to the bathroom\.   \*\-\-\-\-\-\-\-\-\-\-\-\-\-  P\. McGraw \-\-\-\-

**Comment by p5pRT** [comment]
### From @davidnicol

On Tue\, Aug 11\, 2009 at 2&#8203;:31 PM\, Eirik Berg Hanssen\<Eirik\-
>>      As an r\-value\, the list assignment operator passes its own
>> r\-value through\.
>
>  What is "its own r\-value"?  If you mean to say the "elements
> produced by the expression on the right hand side of the assignment"\,
> it is incorrect\.
>
>  \(And if you meant something else\, I found it less clear than what
> the patch had\.\)

Yes\, that's what I meant\.  Thanks for straightening out my misconception\.


>
> use 5\.010;
> use strict;
> use warnings;
> my @&#8203;x = my \($x\, undef\, $y\) = 1\.\.10;
> say $\_||'undef' for @&#8203;x;
> \_\_END\_\_
> 1
> 2
> 3
>
>  3 elements\, not 10\.  Specifically\, the three elements assigned to\.

wow\.

/tmp&#8203;: $ perl \-le '@&#8203;a=qw/24 57 32 hike\!/; $count = \(\) = @&#8203;a; print $count'
4
/tmp&#8203;: $ perl \-le '@&#8203;a=qw/24 57 32 hike\!/; @&#8203;copy = \(\) = @&#8203;a; print @&#8203;copy'

/tmp&#8203;: $

So list assignment in scalar context returns the count of the RHS\, but
list assignment in list context returns the LHS\.



>
>  Aside&#8203;:
>
>  And none of them undef either\.  In fact\, this is one \.\.\. rather
> \_special\_ \.\.\. corner case&#8203;:
>
> use 5\.010;
> use strict;
> use warnings;
> for my $u \(undef\) \{
>  my @&#8203;x = \($u\) = 1\.\.10;
>  say $\_||'undef' for @&#8203;x\, $u;
> \}
> \_\_END\_\_
> 1
> undef
>
>  The assignment to \($u\) evaluates to a list of one element\, the value
> of which differs from the value of $u\, which remains the same both
> before and after the assignment\.  Ah\, the magic of undef\. :\)

Is that not a bug\, that an assignment to a constant \($u is an alias to
undef\) did not cause an exception?  It looks like undef is special
that way\.

/tmp&#8203;: $ perl \-le ' $count = undef = 3; print $count'
Modification of a read\-only value attempted at \-e line 1\.
/tmp&#8203;: $ perl \-le ' \($count\) = \(undef\) = \(3\); print $count'
3
/tmp&#8203;: $ perl \-le ' \($count\) = \(22\) = \(3\); print $count'
Can't modify constant item in list assignment at \-e line 1\, near "\);"
Execution of \-e aborted due to compilation errors\.
/tmp&#8203;: $

**Comment by p5pRT** [comment]
### From @gbarr


On Aug 11\, 2009\, at 3&#8203;:40 PM\, David Nicol wrote&#8203;:
>
>
> /tmp&#8203;: $ perl \-le '@&#8203;a=qw/24 57 32 hike\!/; $count = \(\) = @&#8203;a; print  
> $count'
> 4
> /tmp&#8203;: $ perl \-le '@&#8203;a=qw/24 57 32 hike\!/; @&#8203;copy = \(\) = @&#8203;a; print @&#8203;copy'
>
> /tmp&#8203;: $
>
> So list assignment in scalar context returns the count of the RHS\, but
> list assignment in list context returns the LHS\.

unless the element in the LHS is undef\, in which case it returns the  
RHS value for that element\, as can be seen in

$ perl \-le '@&#8203;a=qw/24 57 32 hike\!/; @&#8203;copy = \($x\,undef\,$y\) = @&#8203;a; $\,="\,";  
print @&#8203;copy'
24\,57\,32

I was about to say that a list assignment in a list context returns  
the first N elements of the RHS\, where N is the number of elements on  
the LHS\, but this is not true when tied variables are involved

\{package Bar;
&nbsp;  sub TIESCALAR \{ return bless \{\} \}
&nbsp;  sub STORE \{ return 99 \}
&nbsp;  sub FETCH \{ return 88 \}
\}
tie $x\, 'Bar';
$\,="\,";
@&#8203;copy = \($x\,$y\) = \(1\,2\);
print @&#8203;copy;

will output

88\,2

Graham\.

**Comment by p5pRT** [comment]
### From Eirik-Berg.Hanssen@allverden.no

Graham Barr \<gbarr@&#8203;pobox\.com> writes&#8203;:

> On Aug 11\, 2009\, at 3&#8203;:40 PM\, David Nicol wrote&#8203;:

>> So list assignment in scalar context returns the count of the RHS\, but
>> list assignment in list context returns the LHS\.
>
> unless the element in the LHS is undef\, in which case it returns the
> RHS value for that element\, as can be seen in
>
> $ perl \-le '@&#8203;a=qw/24 57 32 hike\!/; @&#8203;copy = \($x\,undef\,$y\) = @&#8203;a; $\,="\,";
> print @&#8203;copy'
> 24\,57\,32

&nbsp;  The RHS value\, perhaps\, but not the RHS element itself\.

&nbsp;  Just like your elegant tied\(\) demonstration\, list assignment in
lvalue list context reveals that what is returned is a copy of the RHS
element – even in the case of a constant undef on the LHS&#8203;:

use 5\.010;
use strict;
use warnings;
my $y = 2;
for my $u \(undef\) \{
&nbsp;  my \($x1\, $x2\);
&nbsp;  for my $x \(\($x1\, undef\, $u\, $x2\)=\($y\, $y\, $y\, $y\)\) \{
&nbsp;    $x\*\*=2;
&nbsp;    $y\+\+;
&nbsp;    say "x&#8203;: $x; y&#8203;: $y\, u&#8203;: " \. \($u//'undef'\);
&nbsp;  \}
\}
\_\_END\_\_
x&#8203;: 4; y&#8203;: 3\, u&#8203;: undef
x&#8203;: 4; y&#8203;: 4\, u&#8203;: undef
x&#8203;: 4; y&#8203;: 5\, u&#8203;: undef
x&#8203;: 4; y&#8203;: 6\, u&#8203;: undef

&nbsp;  Compare&#8203;:

use 5\.010;
use strict;
use warnings;
my $y = 2;
for my $x \(\($y\, $y\, $y\, $y\)\) \{
&nbsp;  $x\*\*=2;
&nbsp;  $y\+\+;
&nbsp;  say "x&#8203;: $x; y&#8203;: $y";
\}
\_\_END\_\_
x&#8203;: 5; y&#8203;: 5
x&#8203;: 26; y&#8203;: 26
x&#8203;: 677; y&#8203;: 677
x&#8203;: 458330; y&#8203;: 458330


&nbsp;  Without spending too much text on corner cases nor misleading the
reader too much\, "the list of lvalues assigned to"\, as in the docs
both before and after the proposed patch\, is about as informative as
it gets on this issue\.  Or so I suspect\.


Eirik
\-\- 
The basic facts are that the rate of decrease of the population growth
rate has been falling for decades\, at an ever increasing rate\.
&nbsp;        \-\-jsnead@&#8203;netcom\.com \(John R\. Snead\)

**Comment by p5pRT** [comment]
### From @ig3

Below is a revision of the patch to perlop\.pod which incorporates
some detail of the handling of undef on LHS of a list assignment\,
 From http&#8203;://www\.perlmonks\.org/?node\_id=791113\.

<details><summary>Inline Patch</summary>

```diff
diff --git a/pod/perlop.pod b/pod/perlop.pod
index 1df9fcb..7120a8c 100644
--- a/pod/perlop.pod
+++ b/pod/perlop.pod
@@ -729,9 +729,26 @@ X<assignment> X<operator, assignment> X<=> X<**=> 

```
</details>
X<+=> X<*
 X\<\<\< \<\<= >>> X\<&&=> X\<\-=> X\</=> X\<|=> X\<\<\< >>= >>> X\<||=> X\<//=> X\<\.=>
 X\<%=> X\<^=> X\<x=>

\-"=" is the ordinary assignment operator\.
\+"=" is the ordinary assignment operator\.  There are two forms of this
\+operator&#8203;: list assignment and scalar assignment\.  If the left operand
\+is an array\, a hash\, an array or hash slice or enclosed in parentheses\,
\+then the assignment is a list assignment\.  Otherwise it is a scalar 
assignment\.

\-Assignment operators work as in C\.  That is\,
\+There are other assignment operators\, as in C\. The following are 
recognized&#8203;:
\+
\+    \*\*=    \+=    \*=    &=    \<\<=    &&=
\+           \-=    /=    |=    >>=    ||=
\+           \.=    %=    ^=           //=
\+                 x=
\+
\+These other assignment operators are all scalar assignment operators\.
\+Although they are grouped by family\, they all have the precedence
\+of assignment\.
\+
\+The scalar assignment operators provide scalar context to both their 
operands\.
\+The left operand must be an lvalue \(meaning you can assign to it\)\.
\+
\+The other assignment operators work as in C\. That is\,

&nbsp;     $a \+= 2;

@&#8203;@&#8203; \-740\,18 \+757\,9 @&#8203;@&#8203; is equivalent to
&nbsp;     $a = $a \+ 2;

 although without duplicating any side effects that dereferencing the lvalue
\-might trigger\, such as from tie\(\)\.  Other assignment operators work 
similarly\.
\-The following are recognized&#8203;:
\-
\-    \*\*=    \+=    \*=    &=    \<\<=    &&=
\-           \-=    /=    |=    >>=    ||=
\-           \.=    %=    ^=           //=
\-                 x=
\-
\-Although these are grouped by family\, they all have the precedence
\-of assignment\.
\+might trigger\, such as from tie\(\)\.

\-Unlike in C\, the scalar assignment operator produces a valid lvalue\.
\+Unlike in C\, the scalar assignment operators produce a valid lvalue\.
 Modifying an assignment is equivalent to doing the assignment and
 then modifying the variable that was assigned to\.  This is useful
 for modifying a copy of something\, like this&#8203;:
@&#8203;@&#8203; \-767\,10 \+775\,29 @&#8203;@&#8203; is equivalent to
&nbsp;     $a \+= 2;
&nbsp;     $a \*= 3;

\-Similarly\, a list assignment in list context produces the list of
\-lvalues assigned to\, and a list assignment in scalar context returns
\-the number of elements produced by the expression on the right hand
\-side of the assignment\.
\+The list assignment operator provides list context to both of its operands\.
\+
\+If the left operand of the list assignment operator is enclosed in 
parentheses\,
\+each element of the enclosed list must be an lvalue or C\<undef>\.
\+
\+In scalar context\, the list assignment operator returns the number of
\+elements produced by the expression on the right hand side of the
\+assignment\.
\+
\+In list context\, the list assignment operator returns the list of
\+lvalues assigned to\. Where the value in the LHS list is C\<undef>
\+the corresponding value from the RHS is returned or\, if there is
\+any value in common between the LHS and RHS\, a copy of the corresponding
\+value from the RHS is returned\. The difference is significant if
\+the returned values are modified \(e\.g\. in a for loop or map\)\.
\+
\+    map \{$\_\+\+\} \(\($x\, undef\) = \($y\, $z\)\) \# increments $x and $z
\+    map \{$\_\+\+\} \(\($x\, undef\) = \($y\, $z\, $x\)\) \# increments $x and a copy 
of $z
\+
\+
\+See L\<List value constructors> in L\<perldata> for examples of list 
assignment\.
\+
\+

 =head2 Comma Operator
 X\<comma> X\<operator\, comma> X\<\,>

**Comment by p5pRT** [comment]
### From @jkeenan

There was a lot of back\-and\-forth and revision of patches in this RT
before discussion petered out about 2\-1/2 years ago\.

Could people evaluate the last patch submitted\, below?

Thank you very much\.
Jim Keenan

On Sun Aug 30 08&#8203;:46&#8203;:10 2009\, iang wrote&#8203;:
> Below is a revision of the patch to perlop\.pod which incorporates
> some detail of the handling of undef on LHS of a list assignment\,
>  From http&#8203;://www\.perlmonks\.org/?node\_id=791113\.
> 
> diff \-\-git a/pod/perlop\.pod b/pod/perlop\.pod
> index 1df9fcb\.\.7120a8c 100644
> \-\-\- a/pod/perlop\.pod
> \+\+\+ b/pod/perlop\.pod
> @&#8203;@&#8203; \-729\,9 \+729\,26 @&#8203;@&#8203; X\<assignment> X\<operator\, assignment> X\<=> X\<\*\*=> 
> X\<\+=> X\<\*
>  X\<\<\< \<\<= >>> X\<&&=> X\<\-=> X\</=> X\<|=> X\<\<\< >>= >>> X\<||=> X\<//=> X\<\.=>
>  X\<%=> X\<^=> X\<x=>
> 
> \-"=" is the ordinary assignment operator\.
> \+"=" is the ordinary assignment operator\.  There are two forms of this
> \+operator&#8203;: list assignment and scalar assignment\.  If the left operand
> \+is an array\, a hash\, an array or hash slice or enclosed in parentheses\,
> \+then the assignment is a list assignment\.  Otherwise it is a scalar 
> assignment\.
> 
> \-Assignment operators work as in C\.  That is\,
> \+There are other assignment operators\, as in C\. The following are 
> recognized&#8203;:
> \+
> \+    \*\*=    \+=    \*=    &=    \<\<=    &&=
> \+           \-=    /=    |=    >>=    ||=
> \+           \.=    %=    ^=           //=
> \+                 x=
> \+
> \+These other assignment operators are all scalar assignment operators\.
> \+Although they are grouped by family\, they all have the precedence
> \+of assignment\.
> \+
> \+The scalar assignment operators provide scalar context to both their 
> operands\.
> \+The left operand must be an lvalue \(meaning you can assign to it\)\.
> \+
> \+The other assignment operators work as in C\. That is\,
> 
>      $a \+= 2;
> 
> @&#8203;@&#8203; \-740\,18 \+757\,9 @&#8203;@&#8203; is equivalent to
>      $a = $a \+ 2;
> 
>  although without duplicating any side effects that dereferencing the
lvalue
> \-might trigger\, such as from tie\(\)\.  Other assignment operators work 
> similarly\.
> \-The following are recognized&#8203;:
> \-
> \-    \*\*=    \+=    \*=    &=    \<\<=    &&=
> \-           \-=    /=    |=    >>=    ||=
> \-           \.=    %=    ^=           //=
> \-                 x=
> \-
> \-Although these are grouped by family\, they all have the precedence
> \-of assignment\.
> \+might trigger\, such as from tie\(\)\.
> 
> \-Unlike in C\, the scalar assignment operator produces a valid lvalue\.
> \+Unlike in C\, the scalar assignment operators produce a valid lvalue\.
>  Modifying an assignment is equivalent to doing the assignment and
>  then modifying the variable that was assigned to\.  This is useful
>  for modifying a copy of something\, like this&#8203;:
> @&#8203;@&#8203; \-767\,10 \+775\,29 @&#8203;@&#8203; is equivalent to
>      $a \+= 2;
>      $a \*= 3;
> 
> \-Similarly\, a list assignment in list context produces the list of
> \-lvalues assigned to\, and a list assignment in scalar context returns
> \-the number of elements produced by the expression on the right hand
> \-side of the assignment\.
> \+The list assignment operator provides list context to both of its
operands\.
> \+
> \+If the left operand of the list assignment operator is enclosed in 
> parentheses\,
> \+each element of the enclosed list must be an lvalue or C\<undef>\.
> \+
> \+In scalar context\, the list assignment operator returns the number of
> \+elements produced by the expression on the right hand side of the
> \+assignment\.
> \+
> \+In list context\, the list assignment operator returns the list of
> \+lvalues assigned to\. Where the value in the LHS list is C\<undef>
> \+the corresponding value from the RHS is returned or\, if there is
> \+any value in common between the LHS and RHS\, a copy of the corresponding
> \+value from the RHS is returned\. The difference is significant if
> \+the returned values are modified \(e\.g\. in a for loop or map\)\.
> \+
> \+    map \{$\_\+\+\} \(\($x\, undef\) = \($y\, $z\)\) \# increments $x and $z
> \+    map \{$\_\+\+\} \(\($x\, undef\) = \($y\, $z\, $x\)\) \# increments $x and a copy 
> of $z
> \+
> \+
> \+See L\<List value constructors> in L\<perldata> for examples of list 
> assignment\.
> \+
> \+
> 
>  =head2 Comma Operator
>  X\<comma> X\<operator\, comma> X\<\,>
>

**Comment by p5pRT** [comment]
### From @cpansprout

On Wed Mar 28 18&#8203;:40&#8203;:43 2012\, jkeenan wrote&#8203;:
> There was a lot of back\-and\-forth and revision of patches in this RT
> before discussion petered out about 2\-1/2 years ago\.
> 
> Could people evaluate the last patch submitted\, below?

I thought the consensus was that that was far too much detail\, and that
the exact details may be subject to change\.

> 
> Thank you very much\.
> Jim Keenan
> 
> On Sun Aug 30 08&#8203;:46&#8203;:10 2009\, iang wrote&#8203;:
> > Below is a revision of the patch to perlop\.pod which incorporates
> > some detail of the handling of undef on LHS of a list assignment\,
> >  From http&#8203;://www\.perlmonks\.org/?node\_id=791113\.
> > 
> > diff \-\-git a/pod/perlop\.pod b/pod/perlop\.pod
> > index 1df9fcb\.\.7120a8c 100644
> > \-\-\- a/pod/perlop\.pod
> > \+\+\+ b/pod/perlop\.pod
> > @&#8203;@&#8203; \-729\,9 \+729\,26 @&#8203;@&#8203; X\<assignment> X\<operator\, assignment> X\<=> X\<\*\*=> 
> > X\<\+=> X\<\*
> >  X\<\<\< \<\<= >>> X\<&&=> X\<\-=> X\</=> X\<|=> X\<\<\< >>= >>> X\<||=> X\<//=> X\<\.=>
> >  X\<%=> X\<^=> X\<x=>
> > 
> > \-"=" is the ordinary assignment operator\.
> > \+"=" is the ordinary assignment operator\.  There are two forms of this
> > \+operator&#8203;: list assignment and scalar assignment\.  If the left operand
> > \+is an array\, a hash\, an array or hash slice or enclosed in parentheses\,
> > \+then the assignment is a list assignment\.  Otherwise it is a scalar 
> > assignment\.
> > 
> > \-Assignment operators work as in C\.  That is\,
> > \+There are other assignment operators\, as in C\. The following are 
> > recognized&#8203;:
> > \+
> > \+    \*\*=    \+=    \*=    &=    \<\<=    &&=
> > \+           \-=    /=    |=    >>=    ||=
> > \+           \.=    %=    ^=           //=
> > \+                 x=
> > \+
> > \+These other assignment operators are all scalar assignment operators\.
> > \+Although they are grouped by family\, they all have the precedence
> > \+of assignment\.
> > \+
> > \+The scalar assignment operators provide scalar context to both their 
> > operands\.
> > \+The left operand must be an lvalue \(meaning you can assign to it\)\.
> > \+
> > \+The other assignment operators work as in C\. That is\,
> > 
> >      $a \+= 2;
> > 
> > @&#8203;@&#8203; \-740\,18 \+757\,9 @&#8203;@&#8203; is equivalent to
> >      $a = $a \+ 2;
> > 
> >  although without duplicating any side effects that dereferencing the
> lvalue
> > \-might trigger\, such as from tie\(\)\.  Other assignment operators work 
> > similarly\.
> > \-The following are recognized&#8203;:
> > \-
> > \-    \*\*=    \+=    \*=    &=    \<\<=    &&=
> > \-           \-=    /=    |=    >>=    ||=
> > \-           \.=    %=    ^=           //=
> > \-                 x=
> > \-
> > \-Although these are grouped by family\, they all have the precedence
> > \-of assignment\.
> > \+might trigger\, such as from tie\(\)\.
> > 
> > \-Unlike in C\, the scalar assignment operator produces a valid lvalue\.
> > \+Unlike in C\, the scalar assignment operators produce a valid lvalue\.
> >  Modifying an assignment is equivalent to doing the assignment and
> >  then modifying the variable that was assigned to\.  This is useful
> >  for modifying a copy of something\, like this&#8203;:
> > @&#8203;@&#8203; \-767\,10 \+775\,29 @&#8203;@&#8203; is equivalent to
> >      $a \+= 2;
> >      $a \*= 3;
> > 
> > \-Similarly\, a list assignment in list context produces the list of
> > \-lvalues assigned to\, and a list assignment in scalar context returns
> > \-the number of elements produced by the expression on the right hand
> > \-side of the assignment\.
> > \+The list assignment operator provides list context to both of its
> operands\.
> > \+
> > \+If the left operand of the list assignment operator is enclosed in 
> > parentheses\,
> > \+each element of the enclosed list must be an lvalue or C\<undef>\.
> > \+
> > \+In scalar context\, the list assignment operator returns the number of
> > \+elements produced by the expression on the right hand side of the
> > \+assignment\.
> > \+
> > \+In list context\, the list assignment operator returns the list of
> > \+lvalues assigned to\. Where the value in the LHS list is C\<undef>
> > \+the corresponding value from the RHS is returned or\, if there is
> > \+any value in common between the LHS and RHS\, a copy of the
corresponding
> > \+value from the RHS is returned\. The difference is significant if
> > \+the returned values are modified \(e\.g\. in a for loop or map\)\.
> > \+
> > \+    map \{$\_\+\+\} \(\($x\, undef\) = \($y\, $z\)\) \# increments $x and $z
> > \+    map \{$\_\+\+\} \(\($x\, undef\) = \($y\, $z\, $x\)\) \# increments $x and a copy 
> > of $z
> > \+
> > \+
> > \+See L\<List value constructors> in L\<perldata> for examples of list 
> > assignment\.
> > \+
> > \+
> > 
> >  =head2 Comma Operator
> >  X\<comma> X\<operator\, comma> X\<\,>
> > 
> 
> 


\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @jkeenan

On Wed Mar 28 22&#8203;:58&#8203;:33 2012\, sprout wrote&#8203;:
> On Wed Mar 28 18&#8203;:40&#8203;:43 2012\, jkeenan wrote&#8203;:
> > There was a lot of back\-and\-forth and revision of patches in this RT
> > before discussion petered out about 2\-1/2 years ago\.
> > 
> > Could people evaluate the last patch submitted\, below?
> 
> I thought the consensus was that that was far too much detail\, and that
> the exact details may be subject to change\.
> 
> > 

Should we take that as a recommendation to reject and close the ticket?


Thank you very much\.
Jim Keenan

**Comment by p5pRT** [comment]
### From @cpansprout

On Thu Apr 05 19&#8203;:09&#8203;:01 2012\, jkeenan wrote&#8203;:
> On Wed Mar 28 22&#8203;:58&#8203;:33 2012\, sprout wrote&#8203;:
> > On Wed Mar 28 18&#8203;:40&#8203;:43 2012\, jkeenan wrote&#8203;:
> > > There was a lot of back\-and\-forth and revision of patches in this RT
> > > before discussion petered out about 2\-1/2 years ago\.
> > > 
> > > Could people evaluate the last patch submitted\, below?
> > 
> > I thought the consensus was that that was far too much detail\, and that
> > the exact details may be subject to change\.
> > 
> > > 
> 
> Should we take that as a recommendation to reject and close the ticket?

I think we should close it\, if simply because it makes it easier to fix
bugs if things are not explained in such excruciating detail\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

On Thu Apr 05 19&#8203;:09&#8203;:01 2012\, jkeenan wrote&#8203;:
> On Wed Mar 28 22&#8203;:58&#8203;:33 2012\, sprout wrote&#8203;:
> > On Wed Mar 28 18&#8203;:40&#8203;:43 2012\, jkeenan wrote&#8203;:
> > > There was a lot of back\-and\-forth and revision of patches in this RT
> > > before discussion petered out about 2\-1/2 years ago\.
> > > 
> > > Could people evaluate the last patch submitted\, below?
> > 
> > I thought the consensus was that that was far too much detail\, and that
> > the exact details may be subject to change\.
> > 
> > > 
> 
> Should we take that as a recommendation to reject and close the ticket?

I think we should close it\, if simply because it makes it easier to fix
bugs if things are not explained in such excruciating detail\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @jkeenan

On Fri Apr 06 06&#8203;:22&#8203;:20 2012\, sprout wrote&#8203;:
> On Thu Apr 05 19&#8203;:09&#8203;:01 2012\, jkeenan wrote&#8203;:
> > 
> > Should we take that as a recommendation to reject and close the ticket?
> 
> I think we should close it\, if simply because it makes it easier to fix
> bugs if things are not explained in such excruciating detail\.
> 

Closing ticket\.

**Comment by p5pRT** [comment]
@jkeenan - Status changed from 'open' to 'rejected'
